File: c/p04.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <string.h>

bool is_palindrome(int i) {
    char s[10];
    sprintf(s, "%d", i);
    for (int i=0,j=strlen(s)-1;j>=0; ++i, --j) {
        if (s[i] != s[j]) {
            return false;
        }
    }
    return true;
}

int main () {
    int biggest = 0;
    for (int i = 999; i > 100; --i) {
        for (int j = 990; j > 100; --j) {
            int prod = i * j;
            if (is_palindrome(prod) && prod > biggest) {
                biggest = prod;
            }

        }
    }
    printf("%d\n", biggest);
    return 0;
}

File: ocaml/p04.ml
open Base
open Stdio

let p04 () =
  let is_palindrome int =
    let str = Int.to_string int in
    String.equal (String.rev str) str
  in
  let rec loop i res =
    let rec loop1 j res =
      let prod = i * j in
      if i > 999 then res
      else if j > 999 then loop (i + 1) res
      else if (is_palindrome prod) && (prod > res) then
        loop1 (j + 1) prod
      else
        loop1 (j +  1) res
    in
    loop1 i res
  in
  printf "%d\n" (loop 900 0)

let () =
  Time.time p04
